internal/auth/authn/k8s.go

```diff
@@ -2,32 +2,82 @@ package authn
 
 import (
 	"context"
+	"encoding/json"
+	"fmt"
+	"io"
 	"net/http"
 	"strings"
 )
 
-type K8sAuthN struct{}
+type K8sAuthN struct {
+	K8sApiUrl string
+}
 
 type ctxKeyAuthHeader string
 
-const K8sTokenKey ctxKeyAuthHeader = "k8s-token"
+const (
+	K8sTokenKey ctxKeyAuthHeader = "k8s-token"
+	AuthHeader  string           = "Authorization"
+)
+
+type OauthServerResponse struct {
+	TokenEndpoint string `json:"token_endpoint"`
+}
+
+func (k8s K8sAuthN) AuthHandler(next http.Handler, w http.ResponseWriter, r *http.Request) {
+	authHeader := r.Header.Get(AuthHeader)
+	if authHeader == "" {
+		w.WriteHeader(http.StatusUnauthorized)
+		return
+	}
+
+	authToken := strings.Split(authHeader, "Bearer ")
+	if len(authToken) != 2 {
+		w.WriteHeader(http.StatusUnauthorized)
+		return
+	}
+
+	ctx := context.WithValue(r.Context(), K8sTokenKey, authToken[1])
+	next.ServeHTTP(w, r.WithContext(ctx))
+}
+
+func (k8s K8sAuthN) ValidateToken(ctx context.Context) (bool, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, fmt.Sprintf("%s/apis/user.openshift.io/v1/users/~", k8s.K8sApiUrl), nil)
+	if err != nil {
+		return false, err
+	}
+
+	k8sTokenVal := ctx.Value(K8sTokenKey)
+	if k8sTokenVal == nil {
+		return false, nil
+	}
+	k8sToken := k8sTokenVal.(string)
+
+	req.Header = map[string][]string{
+		"Authorization": {"Bearer " + k8sToken},
+		"Content-Type":  {"application/json"},
+	}
 
-func (k8s K8sAuthN) AuthHandler(next http.Handler) http.Handler {
-	fn := func(w http.ResponseWriter, r *http.Request) {
-		authHeader := r.Header.Get("Authorization")
-		if authHeader == "" {
-			w.WriteHeader(http.StatusUnauthorized)
-			return
-		}
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return false, err
+	}
 
-		authToken := strings.Split(authHeader, "Bearer ")
-		if len(authToken) != 2 {
-			w.WriteHeader(http.StatusUnauthorized)
-			return
-		}
+	return res.StatusCode == http.StatusOK, nil
+}
 
-		ctx := context.WithValue(r.Context(), K8sTokenKey, authToken[1])
-		next.ServeHTTP(w, r.WithContext(ctx))
+func (k8s K8sAuthN) GetTokenRequestURL(ctx context.Context) (string, error) {
+	res, err := http.Get(fmt.Sprintf("%s/.well-known/oauth-authorization-server", k8s.K8sApiUrl))
+	if err != nil {
+		return "", err
+	}
+	oauthResponse := OauthServerResponse{}
+	bodyBytes, err := io.ReadAll(res.Body)
```

Comment:
do we need to close the body after reading?

Reply:
yes, we do! Fixed